sdk/eventhub/Azure.Messaging.EventHubs/tests/Producer/EventHubProducerClientTests.cs

```diff
@@ -645,6 +646,525 @@ public async Task CloseAsyncDoesNotCloseTheConnectionWhenNotOwned()
+                .Throws(new EventHubsException(false, "test", EventHubsException.FailureReason.ClientClosed))
+                .Verifiable();
+
+            transportProducer
+                .SetupGet(transportProducer => transportProducer.IsClosed)
+                .Returns(true);
+
+            SetIsClosed(producerClient, true);
+
+            Assert.That(async () => await producerClient.SendAsync(batch), Throws.InstanceOf<EventHubsException>().And.Property(nameof(EventHubsException.Reason)).EqualTo(EventHubsException.FailureReason.ClientClosed));
+
+            transportProducer.Verify(t => t.SendAsync(It.IsAny<EventDataBatch>(),
+                                                      It.IsAny<CancellationToken>()),
+                                     Times.Once);
+        }
+
+        /// <summary>
+        ///   Retry logic would stop after a cancellation is requested.
+        /// </summary>
+        ///
+        [Test]
+        public void RetryLogicStopsWhenCancellationTriggered()
+        {
+            var options = new SendEventOptions { PartitionId = "0" };
+            var events = new EventData[0];
+            var transportProducer = new Mock<TransportProducer>();
+            var eventHubConnection = new MockConnection(() => transportProducer.Object);
+            var retryPolicy = new EventHubProducerClientOptions().RetryOptions.ToRetryPolicy();
+            var mockTransportProducerPool = new MockTransportProducerPool(transportProducer.Object, eventHubConnection, retryPolicy);
+            var mockPooledProducer = mockTransportProducerPool.GetPooledProducer(options.PartitionId) as MockPooledProducer;
+            var producerClient = new EventHubProducerClient(eventHubConnection, transportProducer.Object, mockTransportProducerPool);
+            var cancellationTokenSource = new CancellationTokenSource();
+            var numberOfCalls = 0;
+            var expectedNumberOfCalls = 2;
+
+            transportProducer
+                .Setup(transportProducer => transportProducer.SendAsync(It.IsAny<IEnumerable<EventData>>(),
+                                                                        It.IsAny<SendEventOptions>(),
+                                                                        It.IsAny<CancellationToken>()))
+                .Callback(() =>
+                {
+                    if (++numberOfCalls >= expectedNumberOfCalls)
+                    {
+                        cancellationTokenSource.Cancel();
+                    }
+
+                    throw new EventHubsException(false, string.Empty, EventHubsException.FailureReason.ClientClosed);
+                })
+                .Returns(Task.CompletedTask);
+
+            transportProducer
+                .SetupGet(transportProducer => transportProducer.IsClosed)
+                .Returns(true);
+
+            Assert.That(async () => await producerClient.SendAsync(events, options, cancellationTokenSource.Token), Throws.Nothing);
```

Comment:
A cancellation should throw a `TaskCanceledException`, shouldn't it? Are we overlooking something.

Reply:
Great point.

We implemented the retry loop similarly to how was done inside the ```AmqpProducer::SendAsync``` method. In there, we only checked that the flag ```cancellationToken.IsCancellationRequested``` was not ```true```.

```csharp
                while (!cancellationToken.IsCancellationRequested)
                {
```

I tried to simplify the test a bit.

Reply:
Thanks for the test simplification.

If you take a look at the `AmqpProducer::SendAsync` method you mentioned, after the main loop there is the following piece of code:

https://github.com/Azure/azure-sdk-for-net/blob/48b622ab7e31fc6091a5719e5e48d7163a25df1d/sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpProducer.cs#L355-L358

We are expected to throw a `TaskCanceledException` to the user when we realize a task cancellation has been requested. This is a way of telling them whether the cancellation was successful or not. [This is actually a common pattern in .NET](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/task-cancellation).

After your changes, if we trigger the task cancellation after the inner `TransportProducer` `SendAsync` method has been called (but before it finishes sending the message), a `TaskCanceledException` could be thrown, and it would be bubbled up to the user (that's expected).

https://github.com/Azure/azure-sdk-for-net/blob/5ffd0376960e3f2b1a5ad4b6193cc49c3edc9d6e/sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClient.cs#L479

However, in some scenarios (like the ones in the tests, where cancellation already took place), we end up leaving the method without throwing the exception. For this reason, including a forced `throw` at the end of the method should keep the behavior consistent. It should throw only if cancellation took place and if the message hasn't been sent.

Reply:
> If you take a look at the `AmqpProducer::SendAsync` method you mentioned, after the main loop there is the following piece of code:
> 
> https://github.com/Azure/azure-sdk-for-net/blob/48b622ab7e31fc6091a5719e5e48d7163a25df1d/sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpProducer.cs#L355-L358

Gotcha! sorry, my bad.

Reply:
Resolving as this is discussed below as well.